adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,248 @@
+            # TODO: compare to ':distributed_tracing' option name/semantics
+            Rack::Adapter.config[:extract_parent_context]
+          end
+
+          def extract_parent_context
+            @parent_context ||= OpenTelemetry::Adapters::Rack::Adapter.propagator.extract(env)
+          end
+
+          def tracer
+            OpenTelemetry::Adapters::Rack::Adapter.tracer
+          end
+
+          ### request_span
+
+          # Sets as many attributes as are available before control
+          # is handed off to next middleware.
+          def start_request_span
+            @request_span ||= tracer.start_span(request_span_name,
+                                                with_parent_context: parent_context,
+                                                # NOTE: try to set as many attributes via 'attributes' argument
+                                                #       instead of via span.set_attribute
+                                                attributes: request_span_attributes,
+                                                kind: :server)
+          end
+
+          # Per https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#http-server-semantic-conventions
+          #
+          # One of the following sets is required, in descending preference:
+          # * http.scheme, http.host, http.target
+          # * http.scheme, http.server_name, net.host.port, http.target
+          # * http.scheme, net.host.name, net.host.port, http.target
+          # * http.url
+          def request_span_attributes
+            {
+              'component' => 'http',
+              'http.method' => env['REQUEST_METHOD'],
+              'http.url' => full_http_request_url,
+              'http.host' => env['HOST'],
+              'http.scheme' => env['rack.url_scheme'],
+              'http.base_url' => base_url # TODO: check attribute naming/semantics
+            }.merge(allowed_request_headers)
+          end
+
+          def full_http_request_url
+            # NOTE: dd-trace-rb has implemented 'quantization'? (which lowers url cardinality)
+            #       see Datadog::Quantization::HTTP.url
+
+            rack_request.url
+          end
+
+          # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#name
+          #
+          def request_span_name
+            # uri_path_value:
+            env['PATH_INFO']
+          end
+
+          def base_url
+            if rack_request.respond_to?(:base_url)
+              rack_request.base_url
+            else
+              # Compatibility for older Rack versions
+              rack_request.url.chomp(rack_request.fullpath)
+            end
+          end
+
+          def rack_request
+            @rack_request ||= ::Rack::Request.new(env)
+          end
+
+          def record_and_reraise_error(error)
+            # TODO: implement span.set_error?
```

Comment:
Currently there are not semantic conventions around errors. I think this is something that needs to be addressed at the specification level. I'm not exactly sure what we should do in the interim. My feeling is that these should probably be represented as events on the span.

Reply:
I set `span.status`, kept a TODO note, kept the rescue clause intact (per SIG discussion).